core/src/main/java/org/elasticsearch/index/query/GeoPolygonQueryBuilder.java

```diff
@@ -67,39 +90,156 @@ public GeoPolygonQueryBuilder addPoint(GeoPoint point) {
         return this;
     }
 
+    public List<GeoPoint> points() {
+        return shell;
+    }
+
     public GeoPolygonQueryBuilder coerce(boolean coerce) {
+        if (coerce) {
+            this.ignoreMalformed = true;
+        }
         this.coerce = coerce;
         return this;
     }
 
+    public boolean coerce() {
+        return this.coerce;
+    }
+
     public GeoPolygonQueryBuilder ignoreMalformed(boolean ignoreMalformed) {
-        this.ignoreMalformed = ignoreMalformed;
+        if (coerce == false) {
+            this.ignoreMalformed = ignoreMalformed;
+        }
         return this;
     }
 
+    public boolean ignoreMalformed() {
+        return ignoreMalformed;
+    }
+
+    @Override
+    public QueryValidationException validate() {
+        QueryValidationException errors = null;
+        if (fieldName == null) {
+            errors = QueryValidationException.addValidationError(NAME, "field must not be null", errors);
+        }
+        if (shell.isEmpty()) {
+            errors = QueryValidationException.addValidationError(NAME, "no points defined for geo_polygon query", errors);
+        } else {
+            GeoPoint start = shell.get(0);
+            if (start.equals(shell.get(shell.size() - 1))) {
+                if (shell.size() < 4) {
+                    errors = QueryValidationException.addValidationError(NAME, "too few points defined for geo_polygon query", errors);
+                }
+            } else {
+                if (shell.size() < 3) {
+                    errors = QueryValidationException.addValidationError(NAME, "too few points defined for geo_polygon query", errors);
+                }
+            }
+        }
+        return errors;
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+
+        if (!shell.get(shell.size() - 1).equals(shell.get(0))) {
+            shell.add(shell.get(0));
+        }
+
+        final boolean indexCreatedBeforeV2_0 = context.parseContext().shardContext().indexVersionCreated().before(Version.V_2_0_0);
+        // validation was not available prior to 2.x, so to support bwc
+        // percolation queries we only ignore_malformed on 2.x created indexes
+        if (!indexCreatedBeforeV2_0 && !ignoreMalformed) {
+            for (GeoPoint point : shell) {
+                if (point.lat() > 90.0 || point.lat() < -90.0) {
```

Comment:
Maybe add a note to switch to https://github.com/elastic/elasticsearch/pull/11969/files#diff-6236367b2788d500f308334adc6d4d7fR70 once that PR is in?